cpp/src/arrow/compute/exec/subtree_internal.h

```diff
@@ -0,0 +1,148 @@
+      expr_to_code_;
+  std::vector<compute::Expression> code_to_expr_;
+  std::unordered_set<expression_codes> subtree_exprs_;
+
+  // Encode a subexpression (returning the existing code if possible).
+  expression_code GetOrInsert(const compute::Expression& expr) {
+    auto next_code = static_cast<int>(expr_to_code_.size());
+    auto it_success = expr_to_code_.emplace(expr, next_code);
+
+    if (it_success.second) {
+      code_to_expr_.push_back(expr);
+    }
+    return it_success.first->second;
+  }
+
+  // Encode an expression (recursively breaking up conjunction members if possible).
+  void EncodeConjunctionMembers(const compute::Expression& expr,
+                                expression_codes* codes) {
+    if (auto call = expr.call()) {
+      if (call->function_name == "and_kleene") {
+        // expr is a conjunction, encode its arguments
+        EncodeConjunctionMembers(call->arguments[0], codes);
+        EncodeConjunctionMembers(call->arguments[1], codes);
+        return;
+      }
+    }
+    // expr is not a conjunction, encode it whole
+    codes->push_back(GetOrInsert(expr));
+  }
+
+  // Convert an encoded subtree or fragment back into an expression.
+  compute::Expression GetSubtreeExpression(const Encoded& encoded_subtree) {
+    // Filters will already be simplified by all of a subtree's ancestors, so
+    // we only need to simplify the filter by the trailing conjunction member
+    // of each subtree.
+    return code_to_expr_[encoded_subtree.partition_expression.back()];
+  }
+
+  // Insert subtrees for each component of an encoded partition expression.
+  void GenerateSubtrees(expression_codes partition_expression,
+                        std::vector<Encoded>* encoded) {
+    while (!partition_expression.empty()) {
+      if (subtree_exprs_.insert(partition_expression).second) {
+        Encoded encoded_subtree{/*fragment_index=*/util::nullopt, partition_expression};
+        encoded->push_back(std::move(encoded_subtree));
+      }
+      partition_expression.resize(partition_expression.size() - 1);
+    }
+  }
+
+  // Encode the fragment's partition expression and generate subtrees for it as well.
+  void EncodeOneExpression(int fragment_index, const Expression& partition_expression,
+                           std::vector<Encoded>* encoded) {
+    Encoded encoded_fragment{fragment_index, {}};
+
+    EncodeConjunctionMembers(partition_expression,
+                             &encoded_fragment.partition_expression);
+
+    GenerateSubtrees(encoded_fragment.partition_expression, encoded);
+
+    encoded->push_back(std::move(encoded_fragment));
+  }
+
+  template <typename Fragments>
+  std::vector<Encoded> EncodeFragments(const Fragments& fragments) {
+    std::vector<Encoded> encoded;
+    for (size_t i = 0; i < fragments.size(); ++i) {
+      EncodeOneExpression(static_cast<int>(i), fragments[i]->partition_expression(),
+                          &encoded);
+    }
+    return encoded;
+  }
```

Comment:
Since this isn't in dataset:: anymore, let's remove references to `Fragment` and make it generically about a set of guarantees
```suggestion
  template <typename GetGuarantee>
  std::vector<Encoded> EncodeGuarantee(const GetGuarantee& get, int count) {
    std::vector<Encoded> encoded;
    for (int i = 0; i < count; ++i) {
      EncodeOneGuarantee(i, get(i), &encoded);
    }
    return encoded;
  }
```